### ISSUE TITLE
Can't get 'section_path_lengths' feature from a population when cache=False

### ISSUE THREAD

**Issue by adrien-berchet** [BlueBrain/NeuroM#926]
I discovered that the new population fails for the `section_path_lengths` feature when `cache=False` . It seems it is because of `node.iupstream()` but I don't get why. Do you have any idea?

In the tests of NeuroM, if I add this in `tests/features/test_get_features.py`, the test fails:
```
def test_section_path_distances_pop():
    section_path_distances = get_feature('section_path_distances', POP)
```
But it works if I set `cache=True` when the population is loaded.
And it seems that only the `section_path_lengths` feature is affected (which is actually not tested on a population in the NeuroM tests).

**Comment by adrien-berchet** [comment]
After some more investigation, I think I found what's happening.

The feature is:
```python
@feature(shape=(...,), name='section_path_distances')
def section_path_lengths(neurites, neurite_type=NeuriteType.all):
    """Path lengths of a collection of neurites."""
    # Calculates and stores the section lengths in one pass,
    # then queries the lengths in the path length iterations.
    # This avoids repeatedly calculating the lengths of the
    # same sections.
    dist = {}
    neurite_filter = is_type(neurite_type)

    for s in iter_sections(neurites, neurite_filter=neurite_filter):
        dist[s] = s.length

    def pl2(node):
        """Calculate the path length using cached section lengths."""
        return sum(dist[n] for n in node.iupstream())

    return _map_sections(pl2, neurites, neurite_type=neurite_type)
```

In `dist`, section addresses are stored as keys. But the morphologies loaded in `iter_sections` are deleted at each iteration. So the sections stored in `dict` are orphans. The simplest solution is probably to just remove the cache for the distance and just do:

```python
@feature(shape=(...,), name='section_path_distances')
def section_path_lengths(neurites, neurite_type=NeuriteType.all):
    """Path lengths of a collection of neurites."""
    def pl2(node):
        """Calculate the path lengt."""
        return sum(n.length for n in node.iupstream())

    return _map_sections(pl2, neurites, neurite_type=neurite_type)
```

It might be less efficient though.

Also, maybe we should add a reference in `morphio.Section` to the morphology that contains it, so it would prevent the garbage collector to drop a morphology whose sections are still in use?

**Comment by arnaudon** [comment]
Is it related to that? 
https://github.com/BlueBrain/MorphIO/issues/161

**Comment by asanin-epfl** [comment]
I see the problem. Thanks for pointing and investigating.

> Also, maybe we should add a reference in morphio.Section to the morphology that contains it, so it would prevent the garbage collector to drop a morphology whose sections are still in use?

This can't be done. See the PR that Alexis pointed out above.

**Comment by asanin-epfl** [comment]
> It might be less efficient though.

I think this is a good example of bad optimizing ahead. I've compared two approaches. The one you suggested is much faster on my machine, and it uses less memory. Always optimize with performance tests! The old approach has two recursions traversals instead of one. The line `dist[s] = s.length` does a heavy operation of calculating section's hash.

```
def section_path_lengths_old(neurites, neurite_type=NeuriteType.all):
    """Path lengths of a collection of neurites."""
    # Calculates and stores the section lengths in one pass,
    # then queries the lengths in the path length iterations.
    # This avoids repeatedly calculating the lengths of the
    # same sections.
    dist = {}
    neurite_filter = is_type(neurite_type)

    for s in iter_sections(neurites, neurite_filter=neurite_filter):
        dist[s] = s.length

    def pl2(node):
        """Calculate the path length using cached section lengths."""
        return sum(dist[n] for n in node.iupstream())

    return _map_sections(pl2, neurites, neurite_type=neurite_type)


def section_path_lengths_new(neurites, neurite_type=NeuriteType.all):
    """Path lengths of a collection of neurites."""

    # Calculates and stores the section lengths in one pass,
    # then queries the lengths in the path length iterations.
    # This avoids repeatedly calculating the lengths of the
    # same sections.
    def pl2(node):
        """Calculate the path length using cached section lengths."""
        return sum(n.length for n in node.iupstream())

    return _map_sections(pl2, neurites, neurite_type=neurite_type)


def cmp_path_lengths():
    nrn = load_neuron('tests/data/neurolucida/bio_neuron-000.asc')
    print('new', timeit.timeit(stmt=partial(section_path_lengths_new, nrn.neurites), number=1000))
    print('old', timeit.timeit(stmt=partial(section_path_lengths_old, nrn.neurites), number=1000))


cmp_path_lengths()
```

**Comment by adrien-berchet** [comment]
I am surprised that the old version is both more complicated and slower... But it's perfect if we can simplify it, make it faster and solve this issue :)

**Comment by adrien-berchet** [comment]
Fixed in #927 
Thanks @asanin-epfl !
